Source/WebKit/UIProcess/mac/_WKCaptionStyleMenuController.mm

```diff
@@ -0,0 +1,148 @@
+    _savedActiveProfileID = CaptionUserPreferencesMediaAF::platformActiveProfileID();
+
+    [self rebuildMenu];
+
+    return self;
+}
+
+- (void)rebuildMenu
+{
+    [_menu removeAllItems];
+    auto profileIDs = CaptionUserPreferencesMediaAF::platformProfileIDs();
+
+    // Add Caption Profile items
+    for (const auto& profileID : profileIDs) {
+        auto title = CaptionUserPreferencesMediaAF::nameForProfileID(profileID);
+        RetainPtr item = adoptNS([[NSMenuItem alloc] initWithTitle:title.createNSString().get() action:@selector(profileMenuItemSelected:) keyEquivalent:@""]);
+        [item setTarget:self];
+        [item setRepresentedObject:profileID.createNSString().get()];
+
+        // Add checkmark for currently selected item
+        if (profileID == _savedActiveProfileID)
+            [item setState:NSControlStateValueOn];
+
+        [_menu addItem:item.get()];
+    }
+
+    if (!profileIDs.isEmpty())
+        [_menu addItem:[NSMenuItem separatorItem]];
+
+    // Add Deep-link to System Caption Settings
+    RetainPtr systemCaptionSettingsItem = adoptNS([[NSMenuItem alloc] initWithTitle:@"Show System Caption Settings" action:@selector(systemCaptionStyleSettingsItemSelected:) keyEquivalent:@""]);
+    [systemCaptionSettingsItem setTarget:self];
+    [systemCaptionSettingsItem setImage:[NSImage imageWithSystemSymbolName:@"gear" accessibilityDescription:nil]];
+    [_menu addItem:systemCaptionSettingsItem.get()];
+}
+
+#pragma mark - Properties
+
+- (NSMenu *)captionStyleMenu
+{
+    return _menu.get();
+}
+
+#pragma mark - Actions
+
+- (void)profileMenuItemSelected:(NSMenuItem *)item
+{
+    if (!item)
+        return;
+
+    NSString *nsProfileID = (NSString *)item.representedObject;
+    if (![nsProfileID isKindOfClass:NSString.class])
+        return;
+
+    _savedActiveProfileID = nsProfileID;
+    CaptionUserPreferencesMediaAF::setActiveProfileID(_savedActiveProfileID);
+    [self rebuildMenu];
+}
+
+- (void)profileMenuItemHighlighted:(NSMenuItem *)item
+{
+    if (!item)
+        return;
+
+    NSString *nsProfileID = (NSString *)item.representedObject;
+    if (![nsProfileID isKindOfClass:NSString.class])
+        return;
+
+    CaptionUserPreferencesMediaAF::setActiveProfileID(nsProfileID);
+}
+
+- (void)systemCaptionStyleSettingsItemSelected:(NSMenuItem *)sender
+{
+    [NSWorkspace.sharedWorkspace openURL:[NSURL URLWithString:@"x-apple.systempreferences:com.apple.Accessibility?Captions"]];
+}
+
+#pragma mark - NSMenuDelegate
+
+- (void)menuWillOpen:(NSMenu *)menu
+{
+    _savedActiveProfileID = CaptionUserPreferencesMediaAF::platformActiveProfileID();
+}
+
+- (void)menu:(NSMenu *)menu willHighlightItem:(NSMenuItem *)item
+{
+    if ([item.representedObject isKindOfClass:NSString.class])
+        [self profileMenuItemHighlighted:item];
+}
+
+- (void)menuDidClose:(NSMenu *)menu
+{
+    if (!_savedActiveProfileID.isEmpty())
+        CaptionUserPreferencesMediaAF::setActiveProfileID(_savedActiveProfileID);
```

Comment:
It might be worth only setting the active profile when `_savedActiveProfileID` changes to avoid spamming the system with a changed notification